### DISCUSSION TITLE
useFieldArray returns never[] if path to name contains optional field

### DISCUSSION THREAD

**Discussion by tparvi** [react-hook-form/react-hook-form#5906] [Q&A]
Note, if you can make the field non optional then this solves the issue: https://github.com/react-hook-form/react-hook-form/discussions/5517

**Unfortunately I cannot change the field to be optional**

Codesandbox: https://codesandbox.io/s/rhf-fieldarray-optional-record-7y4ol?file=/src/App.tsx

In my case the form data contains optional languages and there is an array for each language

```
type Language = "fi" | "en"; // in reality might have +20 languages

type Value = {
  value: string;
};

type ConnectionData = {
  languageVersions: { [lang in Language]?: Value[] }; // each language is optional
};

```

When I use `useFieldArray` and try to construct the name during runtime the fields returned as `never[]` (codesandbox says `any`)

```
  const { fields } = useFieldArray({
    name: `languageVersions.${props.language}`,
    control: props.control
  });
```

This is what it looks like in VS Code with TS 4.3.5
![image](https://user-images.githubusercontent.com/871053/125271966-f937ce80-e313-11eb-9cb2-7b56c44597c4.png)

**Comment by darylf** [comment]
Did you find a solution to this? I feel like I'm experiencing this issue while using a generated Variable type from Graphql-codegen.

    export type UpdateThingMutationVariables = Exact<{
        id: Scalars["ID"];
        credits?: Maybe<Array<CreditInput> | CreditInput>;
        format: Scalars["String"];
        summary?: Maybe<Scalars["String"]>;
    }>;


    export type CreditInput = {
        creatorId: Scalars["ID"];
        featured?: Maybe<Scalars["Boolean"]>;
        roleId: Scalars["ID"];
    };

    const { control } = useForm<UpdateThingMutationVariables>();

    const {fields} = useFieldArray({ control, name: "credits" });

fields is defined as nothing[]

**Reply by tparvi** [reply]
Unfortunately I didn't find any solution to this. I changed all my optional to non-optional as I was running into same problem elsewhere with RHF. So instead of:

`languageVersions: { [lang in Language]?: Value[] }; // each language is optional`

I have:

`languageVersions: { [lang in Language]: Value[] }; // each language is required`

Fortunately in my case I know which languages actually exist so I can take that into account.

**Reply by darylf** [reply]
Thanks! I created an issue for this scenario, as it would be nice if optional arrays could be handled better and referenced this discussion

https://github.com/react-hook-form/react-hook-form/issues/6505
